design/20220720-per-certificate-owner-ref.md

```diff
@@ -0,0 +1,263 @@
+- [Summary](#summary)
+- [Use-cases](#use-cases)
+- [Questions](#questions)
+- [Proposal](#proposal)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Supported Versions](#supported-versions)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+This checklist contains actions which must be completed before a PR implementing this design can be merged.
+
+- [ ] This design doc has been discussed and approved
+- [ ] Test plan has been agreed upon and the tests implemented
+- [ ] Feature gate status has been agreed upon (whether the new functionality will be placed behind a feature gate or not)
+- [ ] Graduation criteria is in place if required (if the new functionality is placed behind a feature gate, how will it graduate between stages)
+- [ ] User-facing documentation has been PR-ed against the release branch in [cert-manager/website](https://github.com/cert-manager/website)
+
+## Summary
+
+The existing flag `--enable-certificate-owner-ref` allows you to configure cert-manager to delete Secret resources when the associated Certificate is removed. 
+
+We propose to introduce the same setting at the Certificate level so that users of the Certificate resource can decide whether the Secret resource should be removed or not.
+
+## Stories
+
+**Story 1: managed cert-manager installations**
+
+[Flant](https://flant.com) manages large multi-tenant Kubernetes clusters. The installation of cert-manager is managed by Flant, and customers cannot edit cert-manager's configuration. Customers have access to a "prod" cluster and a "dev" cluster. On both clusters, Flant uses `--enable-certificate-owner-ref=false` to lower the chance of outages of their managed components such as the ingress controller.
+
+On the "dev" cluster, customers are given long-lived namespaces in which they install and uninstall their applications over and over with random names, including Certificate resources. With hundreds of customers deploying approximately ten times a day to the "dev" cluster, the Secret resources that are left over by cert-manager accumulate (around 10,000 Secret resources after a few months), and the Kubernetes API becomes slow, with people having to wait for 10 seconds to list the secrets in a given namespace.
+
+To solve this problem, Flant aims at using `cleanupPolicy: Never` on the certificates used for their managed components and use `--default-certificate-cleanup-policy=OnDelete` for the rest of the Certificates. Users won't have to change their Certificate resources.
+
+On the "prod" cluster, Flant recommends customers to keep the Secret resource on removal to lower the risk of outages. Flant aims to use `--default-certificate-cleanup-policy=Never` for the "prod" cluster and also aims to document the reason for this difference between "prod" and "dev".
```

Comment:
@wallrj PTAL

Reply:
I can see that the revised story explains why it will not be necessary for the "Customers" to add the cleanupPolicy to each of their Certificate resources.

But my original comment still stands. 

1. This proposed new feature does not address the general problem of "Customers are able to create more resources than the platform API server can handle".
2. Customers could still accidentally bring down the Flant API service by accidentally (or deliberately) creating too many resources.
  1.  Why not instead add a resource count quota?
4. The mission critical Secret resources in the Flant "dev" cluster could still be accidentally deleted.
  1. Wouldn't it be better to implement a solid /  standard mechanism for preventing this situation. (see https://www.linkedin.com/pulse/kubernetes-protecting-resources-from-deletion-them-rodrigo)
5. "Customers" will now see inconsistent behaviour between their "dev" and "production" deployments, in dev their Secrets will be garbage collected and in Prod they will be orphaned. How will they learn about this?


These are only my opinions, and you might find other cert-manager maintainers who are more sympathetic to this new feature.

@maelvls What do you think? Is this how you would solve Flant's problem if you were DevOps there?